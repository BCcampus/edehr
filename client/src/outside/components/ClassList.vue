<template lang="pug">
  div(id="activityList", class="activity-list")
    div(class="activity-list-header columns", v-on:click="activateActivity")
      div(class="header-column is-10 column")
        h3(:title="activity._id") {{ activity.resource_link_title }}
        p LMS description: {{ activity.resource_link_description }}
        p Assignment name: {{ assignment.name }} (LMS configuration: assignment={{ assignment.externalId }} )
        p Assignment description: {{ assignment.description }}
      div(class="header-column is-2 column")
        div(class="header-item header-icon") {{ indicator }}
    div(class="activity-list-body")
      accordion-element(theme="grayTheme", :show="show")
        div(class="classlist-header")
          div(class="classlist-header-item") Evaluation notes
            fas-icon(class="icon-right", icon="download")
        div(class="classlist-body")
          table.table
            thead
              tr
                th Student
                th Submitted
                th Evaluation notes
                th Status
                th
            tbody
              tr(v-for="sv in classList", v-on:click="changeStudent(sv)")
                td
                  div(:id="`ref-${sv._id}`",  :ref="`ref-${sv._id}`") {{ sv.user.fullName }}
                td {{ lastUpdate(sv) }}
                td {{ sv.activityData.evaluationData }}
                td {{ statusText(sv) }}
                td.actions
                  span(v-if="sv.activityData.submitted && !sv.activityData.evaluated")
                    span &nbsp;
                    ui-button(v-on:buttonClicked="unsubmit(sv)", v-bind:secondary="true", :title="unsubmitTool") {{unsubmitText}}
                  span(v-if="sv.activityData.submitted && !sv.activityData.evaluated")
                    ui-button(v-on:buttonClicked="goToEhr(sv)", v-bind:secondary="true", title="View and evaluate in the EHR") Evaluate student work
                  span(v-if="showEvaluateAction(sv)") &nbsp;
                    ui-button(v-on:buttonClicked="markEvaluated(sv)", v-bind:secondary="true", :title="evaluatedButtonTooltip(sv)") {{ evaluatedButtonText(sv) }}

</template>

<script>
import AccordionElement from '../../app/components/AccordionElement'
import UiButton from '../../app/ui/UiButton.vue'
import { formatTimeStr } from '../../helpers/ehr-utills'
import StoreHelper from '../../helpers/store-helper'

export default {
  name: 'ActivityList',
  components: {
    AccordionElement,
    UiButton
  },
  data () {
    return {
      show: false,
      indicator: '+',
      unsubmitText: 'Send back for edits',
      unsubmitTool: 'Send back for edits',
      classList: []
    }
  },
  props: {
    activity: { type: Object },
    index: { type: Number}
  },
  computed: {
    assignment () {
      return this.activity.assignment || {}
    },
    visitInfo () {
      return this.$store.state.visit.sVisitInfo
    }
  },
  methods: {
    evaluatedButtonText (sv) {
      return sv.activityData.evaluated ? 'Take back from student' : 'Send evaluation to student'
    },
    evaluatedButtonTooltip (sv) {
      return sv.activityData.evaluated ? 'I want to edit the evaluation notes' : 'Evaluation is done. Let the student see the evaluation notes.'
    },
    showEvaluateLabel (sv) {
      return sv.activityData.evaluated
    },
    showEvaluateAction (sv) {
      return sv.activityData.submitted && sv.activityData.evaluationData
    },
    setShow: function (value) {
      this.show = value
      this.indicator = value ? '-' : '+'
    },
    statusText (sv) {
      let result = sv.activityData.submitted ? 'Submitted and waiting for evaluation' : 'Not submitted'
      
      result = sv.activityData.evaluated ? 'Evaluated' : result
      return result
    },
    activateActivity () {
      if (this.show) {
        this.setShow(false)
        return
      }
      let activityId = this.activity._id
      localStorage.setItem('activityId', activityId)
      this.loadActivity(activityId)
    },
    loadActivity (activityId) {
      const _this = this
      return StoreHelper.dispatchLoadActivity(this, activityId)
        .then(() => {
          return StoreHelper.dispatchLoadClassList(this, activityId)
        })
        .then((classList) => {
          console.log('ClassList have classlist ', classList)
          _this.classList = classList
          _this.$nextTick(function () {
            _this.setShow(true)
          })
        })
    },
    unsubmit (sv) {
      let activityDataId = sv.activityData._id
      let data = { activityDataId: activityDataId, value: false }
      this.$store.dispatch('ehrData/sendSubmitOverride', data)
        .then((activityData) => {
          let options = {sv: sv, activityData: activityData}
          this.$store.commit('instructor/updateActivityData', options)
        })
    },
    markEvaluated (sv) {
      const newState = ! sv.activityData.evaluated
      let data = { activityDataId: sv.activityData._id, evaluated: newState }
      this.$store.dispatch('ehrData/sendEvaluated', data)
        .then((activityData) => {
          let options = {sv: sv, activityData: activityData}
          this.$store.commit('instructor/updateActivityData', options)
        })
    },
    lastUpdate (sv) {
      return formatTimeStr(sv.activityData.lastDate)
    },
    goToEhr (studentVisit) {
      let studentId = studentVisit._id
      // Go to the first screen related to the assignent
      let name = '/ehr/patient/demographics'
      StoreHelper.setIsDevelopingContent(this,false)
      StoreHelper.setInstructorReturnUrl(this, window.location.pathname)
      // console.log('Store the current student id that is being evaluated ', studentId)
      StoreHelper.dispatchChangeCurrentEvaluationStudentId(this, this.classList, studentId).then(() => {
        console.log('go to ehr with ', name)
        this.$router.push(name)
      })
    },
    changeStudent (sv) {
      let pid = sv._id
      StoreHelper.dispatchChangeCurrentEvaluationStudentId(this, this.classList, pid)
    }
  },
  mounted: function () {
    if (this.index === 0) {
      const _this = this
      this.$nextTick(function () {
        _this.activateActivity()
      })

    }
    /*
    let myId = this.activity._id
    let storeId = localStorage.getItem('activityId')
    if (storeId === myId) {
      // TODO  see notes in AsInstructor view regaring scrolling
      this.loadActivity(storeId).then(() => {
        let list = this.classList
        if (list && list.length > 0) {
          let lastStudent = list[list.length - 1]
          let id = `ref-${lastStudent._id}`
          const _this = this
          this.$nextTick(function() {
            // console.log('Look for element with id', id)
            let elem = document.getElementById(id)
            // console.log('scroll to ', elem)
            _this.$scrollTo(elem)
          })
        }
      })
    }
    */
  }
}
</script>

<style lang="scss" scoped>
@import '../../scss/definitions';
.activity-list {
  padding: 0;
}
.activity-list-header {
  background-color: $grey03;
  border: 1px solid rgba(255, 255, 255, 0);
  margin-bottom: 0;
  cursor: pointer;

  .header-column {
    padding: 1rem 1.5rem;

    p {
      margin-bottom: .5rem;
    }
  }
  .header-item {
    display: block;
  }
  .header-icon {
    font-size: 2rem;
    font-weight: bold;
    text-align: right;
  }
}

.activity-list-body {
  background-color: $grey10;
  overflow: hidden;
  margin-bottom: 0;

  .classlist-header {
    padding: 0.5rem 1.5rem;
    background-color: $grey10;
    border: 1px solid $grey20;
    box-sizing: border-box;
  }
  .classlist-header-item {
    display: inline-block;
    margin-right: 2rem;
  }
  .classlist-body {
    padding: 1rem 1.5rem;
    background-color: $white;
    border: 1px solid $grey20;
    box-sizing: border-box;
    overflow: hidden;
  }
  .table {
    margin-bottom: 0;
    overflow: hidden;
    width: 100%;
    button {
      margin-bottom: 0;
    }
    tr:last-child {
      border-bottom: 0;
    }
    td.actions {
      text-align: right;
      & button {
        margin-left: 20px;
      }
    }
  }
}
.activity-list:hover .activity-list-header {
  background-color: $brand-primary-light;
  box-sizing: border-box;
  border: 1px solid $brand-primary;
}
</style>
