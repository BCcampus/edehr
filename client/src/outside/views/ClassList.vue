<template lang="pug">
  div
    class-list
</template>

<script>
import ClassList from '../components/ClassList'

export default {
  components: {
    ClassList
  }
}
</script>
