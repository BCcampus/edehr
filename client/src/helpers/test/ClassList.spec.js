import { shallowMount, createLocalVue } from '@vue/test-utils'
import ClassList from '../../outside/components/ClassList.vue'
import VueRouter from 'vue-router'
import { library } from '@fortawesome/fontawesome-svg-core'
import {
  faDownload,
} from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
library.add(faDownload)

const localVue = createLocalVue()

const router = new VueRouter()
localVue.use(VueRouter)
localVue.component('fas-icon', FontAwesomeIcon)

const wrapper =  shallowMount(ClassList, {
  localVue,
  router
})


describe('ClassName test', () => {
  // let wrapper

  describe('Basic testing', () => {
    test('Mounts properly', () =>{
      expect(wrapper.isVueInstance()).toBeTruthy()
    })
    test('renders as expected', () => {
      expect(wrapper.html()).toMatchSnapshot()
    })
    test('has "Download Evaluation Notes as CSV button"', () => {
      const button = wrapper.find('.assignmentascsv')
      expect(button.text()).toContain(
        'Download all assignment evaluation notes'
      )
    })
  })
 


})