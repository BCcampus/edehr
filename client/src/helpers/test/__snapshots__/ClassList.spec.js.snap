// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClassName test Basic testing renders as expected 1`] = `
<div>
  <div class="classlist-header-group">
    <div class="course-header">
      <h2 class="course-header-item"><span>: &nbsp;</span><a href="/assignments" class="is-link"><span></span></a></h2>
    </div>
    <div class="course-header-item float-right"><button class="button is-primary" title="Block students from doing more work. Instructor can evaluate all work.">Close activity<svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="hourglass-end" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 384 512" class="icon-right svg-inline--fa fa-hourglass-end fa-w-12">
          <path fill="currentColor" d="M360 64c13.255 0 24-10.745 24-24V24c0-13.255-10.745-24-24-24H24C10.745 0 0 10.745 0 24v16c0 13.255 10.745 24 24 24 0 90.965 51.016 167.734 120.842 192C75.016 280.266 24 357.035 24 448c-13.255 0-24 10.745-24 24v16c0 13.255 10.745 24 24 24h336c13.255 0 24-10.745 24-24v-16c0-13.255-10.745-24-24-24 0-90.965-51.016-167.734-120.842-192C308.984 231.734 360 154.965 360 64zM192 208c-57.787 0-104-66.518-104-144h208c0 77.945-46.51 144-104 144z" class=""></path>
        </svg></button><button class="button is-primary" data-test-id="ClassList.button.download">Download all assignment evaluation notes<svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="download" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512" class="icon-right svg-inline--fa fa-download fa-w-16">
          <path fill="currentColor" d="M216 0h80c13.3 0 24 10.7 24 24v168h87.7c17.8 0 26.7 21.5 14.1 34.1L269.7 378.3c-7.5 7.5-19.8 7.5-27.3 0L90.1 226.1c-12.6-12.6-3.7-34.1 14.1-34.1H192V24c0-13.3 10.7-24 24-24zm296 376v112c0 13.3-10.7 24-24 24H24c-13.3 0-24-10.7-24-24V376c0-13.3 10.7-24 24-24h146.7l49 49c20.1 20.1 52.5 20.1 72.6 0l49-49H488c13.3 0 24 10.7 24 24zm-124 88c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20zm64 0c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20z" class=""></path>
        </svg></button>
      <div data-test-id="ClassList.prompt.saveAs">
        <div style="display: none;" name="dialog">
          <div class="modal-mask"></div>
          <div class="dialog-wrapper" style="top: 0px; left: 0px;">
            <div class="dialog-move-bar"></div>
            <div class="dialog-header">
              <h2>Save evaluation</h2><button class="button button-close"><svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="times" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 352 512" class="icon-right svg-inline--fa fa-times fa-w-11">
                  <path fill="currentColor" d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z" class=""></path>
                </svg></button>
            </div>
            <div class="dialog-body">
              <div>
                <div class="instruction">Save evaluations for undefined</div>
                <div class="ehr-group-wrapper grid-left-to-right-2">
                  <div class="form-element">
                    <div class="text_input_wrapper"><input type="text" class="input text-input" data-focus="false"><span class="suffix">.csv</span></div>
                  </div>
                </div>
              </div>
            </div>
            <div class="dialog-footer">
              <div class="dialog-footer-errors is-pulled-left">
                <div style="display: none;"><span>Correct the following: </span><span class="error-color"><span></span></span></div>
              </div>
              <div class="dialog-footer-content">
                <!---->
                <div class="is-pulled-right"><button class="button is-light">Cancel</button>
                  <div class="dialog-footer-button-space"></div><button disabled="disabled" class="button is-primary">OK</button>
                </div>
              </div>
            </div>
          </div>
          <!---->
        </div>
      </div>
    </div>
  </div>
  <div></div>
  <table>
    <tr>
      <td>Last Update:</td>
      <td></td>
    </tr>
    <tr>
      <td>Created:</td>
      <td></td>
    </tr>
    <tr>
      <td>Status:</td>
      <td>Open</td>
    </tr>
    <!---->
    <tr>
      <td>Students participating:</td>
      <td>0</td>
    </tr>
    <tr>
      <td>Students submitted:</td>
      <td>0</td>
    </tr>
  </table>
  <div class="classlist-body">
    <table class="table">
      <thead>
        <tr>
          <th>Student</th>
          <th>Submitted</th>
          <th>Evaluation notes</th>
          <th>Status</th>
          <th></th>
        </tr>
      </thead>
      <tbody></tbody>
    </table>
  </div>
</div>
`;
